### ISSUE TITLE
Promise.all resolves early when a promise is chained

### ISSUE THREAD

**Issue by atomanyih** [charleshansen/mock-promises#22]
we ran into this bug:

```js
describe('this bug', () => {
  it('this works fine', () => {
    const whySpy = jasmine.createSpy('whyyy');

    const promise1 = new Deferred();
    const promise2 = new Deferred();

    Promise.all([promise1, promise2]).then(whySpy);

    promise1.resolve();
    MockPromises.tick(10);

    expect(whySpy).not.toHaveBeenCalled();
  });

  it('this fails', () => {
    const whySpy = jasmine.createSpy('whyyy');

    const promise1 = new Deferred();
    const promise2 = new Deferred();
    const promise3 = new Deferred();

    const promise1Chain = promise1.then(() => {
      return promise3
    });

    Promise.all([promise1Chain, promise2]).then(whySpy);

    promise1.resolve();
    promise3.resolve();
    MockPromises.tick(10);

    expect(whySpy).not.toHaveBeenCalled();
  });
});
```

second case shouldn't resolve until `promise2` has resolved but it appears `promise1` and `promise3` both count towards `numResolved` in `MockPromises.fakeAll` and it calls whySpy early

@clairethompson

**Comment by atomanyih** [comment]
further investigation: 

```js
it('also fails', () => {
    const whySpy = jasmine.createSpy('whyyy');

    const promise1 = new Deferred();
    const promise3 = new Deferred();

    promise1.then(() => {
      return promise3
    }).then(whySpy);

    promise1.resolve();
    promise3.resolve();

    expect(whySpy).not.toHaveBeenCalled();
    MockPromises.tick(10);
    expect(whySpy).toHaveBeenCalled();
  });
```

we are using native promises

**Comment by rdy** [comment]
this isn't a bug, in the implementation of your deferred object by default calling resolve automatically "ticks" that promise. If you don't want this behavior I'd probably create a new Promise instead of using the deferred, its a convenience over normal promises.
```

**Comment by atomanyih** [comment]
Ok, that explains the deal w my second comment, so disregard that. The bug
I’m trying to report is the second test case in the first comment. Unless
I’ve totally misunderstood Promise.all it shouldn’t resolve until promise2
is resolved no matter how much ticking deferred does

**Comment by charleshansen** [comment]
I tried adding your test to the mock-promises test suite and it already passes. How are you mocking the native promises?

**Comment by charleshansen** [comment]
Never mind, I was able to reproduce the failing test. It looks like `executeForPromise` is a little aggressive here. Not sure how to fix it yet.

**Comment by charleshansen** [comment]
I just fixed it in 0.6.2. Let me know if it works for you.

It turns out there was a bug where sometimes if you have a then callback attached to a promise that resolved to an unresolved promise, the callback would be called twice. This was messing with the internal accounting in Promise.all.
